### PR TITLE
allow managed clusters for ci-entrypoint

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR makes it possible to use ci-entrypoint.sh for clusters representing AKS clusters and using the ASO API by making some small changes. I'll add more details inline.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/create-identity-secret.sh [hunk 1]
```diff
@@ -28,8 +28,17 @@ make --directory="${REPO_ROOT}" "${KUBECTL##*/}"
 # shellcheck source=hack/parse-prow-creds.sh
 source "${REPO_ROOT}/hack/parse-prow-creds.sh"
 
-if ! "${KUBECTL}" get secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" > /dev/null; then
+if ! "${KUBECTL}" get secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" > /dev/null && [[ -v "AZURE_CLIENT_SECRET" ]]; then
```
Comment (jackfrancis) [inline hack/create-identity-secret.sh:31]: why "-v" and not "-n" ?

File: hack/create-identity-secret.sh [hunk 2]
```diff
@@ -28,8 +28,17 @@ make --directory="${REPO_ROOT}" "${KUBECTL##*/}"
...
   "${KUBECTL}" create secret generic "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
 fi
 
-"${KUBECTL}" label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
+if "${KUBECTL}" get secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" > /dev/null; then
```
Comment (jackfrancis) [inline hack/create-identity-secret.sh:35]: this probably fixes some flakes

Comment (jackfrancis) [reply hack/create-identity-secret.sh:35]: actually should this be an `until`?

File: hack/create-identity-secret.sh (hunk 1 above)
Comment (nojnhuh) [reply hack/create-identity-secret.sh:31]: I think both would be practically equivalent here. From bash(1):
```
-v varname
              True if the shell variable varname is set (has been assigned a value).
-n string
              True if the length of string is non-zero.
```

File: hack/create-identity-secret.sh [hunk 3]
```diff
@@ -28,8 +28,17 @@ make --directory="${REPO_ROOT}" "${KUBECTL##*/}"
...
+  "${KUBECTL}" label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
+fi
+
+if ! "${KUBECTL}" get secret "${ASO_CREDENTIAL_SECRET_NAME}" > /dev/null; then
+  "${KUBECTL}" create secret generic "${ASO_CREDENTIAL_SECRET_NAME}" --from-literal=AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}" --from-literal=AZURE_TENANT_ID="${AZURE_TENANT_ID}" --from-literal=AZURE_CLIENT_ID="${AZURE_CLIENT_ID}" --from-literal=AUTH_MODE="${ASO_CREDENTIAL_SECRET_MODE:-workloadidentity}"
+fi
+
```
Comment (jackfrancis) [inline hack/create-identity-secret.sh:42]: should we add an `until` here as well before the label operation to account for the flows where we just created the secret a few ticks ago?

File: scripts/ci-entrypoint.sh [hunk 4]
```diff
@@ -250,8 +248,16 @@ export ARTIFACTS="${ARTIFACTS:-${PWD}/_artifacts}"
 # create cluster
 create_cluster
 
-# install CNI and CCM
-install_addons
+# export the target cluster KUBECONFIG if not already set
+export KUBECONFIG="${KUBECONFIG:-${PWD}/kubeconfig}"
```
Comment (jackfrancis) [inline scripts/ci-entrypoint.sh:252]: is this net new? if so how is this working for the similarly implemented cloud-provider-azure tests that rely upon ci-entrypoint.sh?

File: hack/create-identity-secret.sh (hunk 1 above)
Comment (jackfrancis) [reply hack/create-identity-secret.sh:31]: right, the edge case is if `$AZURE_CLIENT_SECRET` is set to an empty string then we are going to make a non-functional secret

this is probably academic, but that's why `-v` is probably not sufficient for what we're actually doing w/ the value of `$AZURE_CLIENT_SECRET`

File: scripts/ci-entrypoint.sh (hunk 4 above)
Comment (nojnhuh) [reply scripts/ci-entrypoint.sh:252]: This is factored out of `install_addons` so it runs even in the cases where `install_addons` no longer will with these changes.

File: hack/create-identity-secret.sh (hunk 3 above)
Comment (nojnhuh) [reply hack/create-identity-secret.sh:42]: I think the API is generally robust enough to correctly handle sequences like this with its own synchronization. Or have you seen a flake like that before?

File: scripts/ci-entrypoint.sh [hunk 5]
```diff
@@ -250,8 +248,16 @@ export ARTIFACTS="${ARTIFACTS:-${PWD}/_artifacts}"
...
+
+if [[ ! "${CLUSTER_TEMPLATE}" =~ "aks" ]]; then
+  # install CNI and CCM
+  install_addons
```
[the remark this answers is not in the document]
Comment (willie-yao) [reply scripts/ci-entrypoint.sh:256]: Yeah looking at the function I'm a bit confused why it's named that way. Should this be renamed for clarity?

**Review by willie-yao — COMMENTED** [review]
Looks good overall!

File: hack/create-identity-secret.sh [hunk 6]
```diff
@@ -28,8 +28,17 @@ make --directory="${REPO_ROOT}" "${KUBECTL##*/}"
...
   "${KUBECTL}" create secret generic "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
 fi
 
-"${KUBECTL}" label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
+if "${KUBECTL}" get secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" > /dev/null; then
+  "${KUBECTL}" label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
+fi
+
+if ! "${KUBECTL}" get secret "${ASO_CREDENTIAL_SECRET_NAME}" > /dev/null; then
+  "${KUBECTL}" create secret generic "${ASO_CREDENTIAL_SECRET_NAME}" --from-literal=AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}" --from-literal=AZURE_TENANT_ID="${AZURE_TENANT_ID}" --from-literal=AZURE_CLIENT_ID="${AZURE_CLIENT_ID}" --from-literal=AUTH_MODE="${ASO_CREDENTIAL_SECRET_MODE:-workloadidentity}"
```
Comment (nojnhuh) [inline hack/create-identity-secret.sh:40]: I think using `AZURE_CLIENT_ID` here is what's causing issues in the test where it should be `AZURE_CLIENT_ID_USER_ASSIGNED_IDENTITY`, but that isn't exported here. Since that is usually used directly from a template in an AzureClusterIdentity, I think I'll revert these changes and instead define this Secret in the template in the cluster-autoscaler repo. That Secret won't contain anything we don't already put in an AzureClusterIdentity we define in other templates, so that seems reasonable to me.

**Comment by nojnhuh** [comment]
we have green! https://github.com/kubernetes/autoscaler/pull/6975

/hold cancel

**Review by willie-yao — APPROVED** [review]
/lgtm

**Comment by nojnhuh** [comment]
> /hold for [kubernetes/autoscaler#6975 (comment)](https://github.com/kubernetes/autoscaler/pull/6975#issuecomment-2192175889)

All green over there (for real), so this is ready for review again. Leaving the hold until I squash.

**Comment by jackfrancis** [comment]
after squash commits of course :)

**Comment by nojnhuh** [comment]
squashed!
/hold cancel
